### PR TITLE
8318447: Move NMT source code to own subdirectory

### PR DESCRIPTION
**Pull request by jdksjolen**
I think that NMT is deserving of its own subdirectory. Can we do a review of the changes before I fix the merge conflicts?

1. Moved all the nmt source code from services/ to nmt/
2. Renamed all the include statements and sorted them
3. Fixed the include guards

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318447](https://bugs.openjdk.org/browse/JDK-8318447): Move NMT source code to own subdirectory (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [4dfb027e](https://git.openjdk.org/jdk/pull/16276/files/4dfb027ec1260f351d13f4c1b013c35ca02e3aa8)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [2fd7c355](https://git.openjdk.org/jdk/pull/16276/files/2fd7c35576584538e2c1ce8682d7845938a66478)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16276/head:pull/16276` \
`$ git checkout pull/16276`

Update a local copy of the PR: \
`$ git checkout pull/16276` \
`$ git pull https://git.openjdk.org/jdk.git pull/16276/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16276`

View PR using the GUI difftool: \
`$ git pr show -t 16276`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16276.diff">https://git.openjdk.org/jdk/pull/16276.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16276#issuecomment-1771649320)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -85,8 +86,7 @@
 #include "runtime/orderAccess.hpp"
 #include "runtime/safepointMechanism.hpp"
 #include "runtime/vmThread.hpp"
-#include "services/mallocTracker.hpp"
-#include "services/memTracker.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:89]: This included end up at the wrong place.

File: src/hotspot/share/nmt/memBaseline.cpp [hunk 2]
```diff
@@ -28,8 +28,8 @@
 #include "memory/metaspaceUtils.hpp"
 #include "runtime/javaThread.hpp"
 #include "runtime/safepoint.hpp"
-#include "services/memBaseline.hpp"
-#include "services/memTracker.hpp"
+#include "nmt/memBaseline.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/memBaseline.cpp:32]: Sort order

File: src/hotspot/share/nmt/nmtPreInit.hpp [hunk 3]
```diff
@@ -30,7 +30,7 @@
 #ifdef ASSERT
 #include "runtime/atomic.hpp"
 #endif
-#include "services/memTracker.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/nmtPreInit.hpp:33]: The ASSERT Section should be moved down to after the main include block, or even better just include it without the guard. Might be nice to fix this while you are moving around the includes.

File: src/hotspot/share/nmt/threadStackTracker.cpp [hunk 4]
```diff
@@ -24,11 +24,11 @@
...
 
+#include "nmt/mallocTracker.hpp"
 #include "runtime/threadCritical.hpp"
-#include "services/mallocTracker.hpp"
-#include "services/memTracker.hpp"
-#include "services/virtualMemoryTracker.hpp"
-#include "services/threadStackTracker.hpp"
+#include "nmt/memTracker.hpp"
+#include "nmt/threadStackTracker.hpp"
+#include "nmt/virtualMemoryTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/threadStackTracker.cpp:31]: Sort order

File: src/hotspot/share/nmt/virtualMemoryTracker.hpp [hunk 5]
```diff
@@ -28,13 +28,12 @@
 #include "memory/allocation.hpp"
 #include "memory/metaspace.hpp" // For MetadataType
 #include "memory/metaspaceStats.hpp"
-#include "services/allocationSite.hpp"
-#include "services/nmtCommon.hpp"
+#include "nmt/allocationSite.hpp"
+#include "nmt/nmtCommon.hpp"
 #include "utilities/linkedlist.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/virtualMemoryTracker.hpp:33]: This file didn't get an update to the include guard

File: test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp [hunk 6]
```diff
@@ -24,12 +24,13 @@
 
 #include "precompiled.hpp"
 #include "memory/allocation.hpp"
+#include "nmt/memTracker.hpp"
 #include "runtime/os.hpp"
-#include "services/memTracker.hpp"
 #include "utilities/debug.hpp"
 #include "utilities/ostream.hpp"
-#include "unittest.hpp"
+
```
Comment (stefank) [inline test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp:31]: You make this change in some files but not all. Maybe revert this unrelated change?

File: test/hotspot/gtest/nmt/test_nmt_cornercases.cpp [hunk 7]
```diff
@@ -23,11 +23,12 @@
  */
 
 #include "precompiled.hpp"
+
```
Comment (stefank) [inline test/hotspot/gtest/nmt/test_nmt_cornercases.cpp:26]: You make this change in some files but not all. Maybe revert this unrelated change?

File: src/hotspot/share/gc/parallel/psParallelCompact.cpp [hunk 8]
```diff
@@ -30,9 +30,9 @@
 #include "classfile/systemDictionary.hpp"
 #include "code/codeCache.hpp"
 #include "compiler/oopMap.hpp"
+#include "gc/parallel/parMarkBitMap.inline.hpp"
```
Comment (stefank) [inline src/hotspot/share/gc/parallel/psParallelCompact.cpp:33]: This uses a case-sensitive sort, whereas I think that when we add includes without using a sorting tool we would do so in a case-insensitive manner. I'm not sure we should make this change here. (The same goes for similar cases in the rest of the patch)

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/share/nmt/nmtPreInit.hpp [hunk 9]
```diff
@@ -30,7 +30,7 @@
 #ifdef ASSERT
 #include "runtime/atomic.hpp"
 #endif
-#include "services/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/nmtPreInit.hpp:33]: Missing include guard rename

File: src/hotspot/os/linux/os_linux.cpp [hunk 10]
```diff
@@ -60,21 +61,20 @@
 #include "runtime/statSampler.hpp"
 #include "runtime/stubRoutines.hpp"
 #include "runtime/threadCritical.hpp"
-#include "runtime/threads.hpp"
 #include "runtime/threadSMR.hpp"
+#include "runtime/threads.hpp"
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:65]: This was correct originally - lowercase sorts before upper.

File: src/hotspot/share/nmt/mallocTracker.inline.hpp [hunk 11]
```diff
@@ -26,8 +26,8 @@
 #ifndef SHARE_SERVICES_MALLOCTRACKER_INLINE_HPP
 #define SHARE_SERVICES_MALLOCTRACKER_INLINE_HPP
 
+#include "nmt/mallocTracker.hpp"
 #include "services/mallocLimit.hpp"
-#include "services/mallocTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/mallocTracker.inline.hpp:30]: Missing include guard rename

**Review by stefank — CHANGES_REQUESTED** [review]

**Review by dholmes-ora — APPROVED** [review]
Looks okay for moving the NMT files.

The unrelated include file ordering changes just made it harder to review though.

Thanks.

File: test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp [hunk 12]
```diff
@@ -23,10 +23,11 @@
  */
 
 #include "precompiled.hpp"
+
 #include "memory/allocation.hpp"
+#include "nmt/mallocHeader.inline.hpp"
 #include "runtime/os.hpp"
-#include "services/mallocHeader.inline.hpp"
-#include "services/memTracker.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp:30]: order

**Review by tstuefe — APPROVED** [review]
Mostly okay. Was overdue.
Don't need another look if you fix Stefan's and my remarks.

**Comment by jdksjolen** [comment]
Hi,

Thank you for looking through these changes. I applied your comments and also did a run through to look for incorrectly ordered includes. For the gtest source files I separated the includes in a consistent manner, they all look like this pattern now:

```c++
#include "precompiled.hpp"
#include "memory/allocation.hpp"
#include "nmt/mallocHeader.inline.hpp"
#include "nmt/memTracker.hpp"
#include "runtime/os.hpp"

#include "testutils.hpp"
#include "unittest.hpp"
```

**Review by dholmes-ora — APPROVED** [review]
Still good.

(Your merge commit seems to have confused GitHub for some reason.)

File: src/hotspot/share/nmt/memTracker.inline.hpp [hunk 13]
```diff
@@ -23,11 +23,11 @@
...
 
-#ifndef SHARE_SERVICES_MEMTRACKER_INLINE_HPP
-#define SHARE_SERVICES_MEMTRACKER_INLINE_HPP
+#ifndef SHARE_NMT_MEMTRACKER_INLINE_HPP
+#define SHARE_NMT_MEMTRACKER_INLINE_HPP
 
-#include "services/mallocTracker.inline.hpp"
-#include "services/memTracker.hpp"
+#include "nmt/mallocTracker.inline.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/memTracker.inline.hpp:30]: Bonus points if you fix the include for this .inline.hpp file.
```suggestion
#include "nmt/memTracker.hpp"

#include "nmt/mallocTracker.inline.hpp"
```

File: src/hotspot/share/nmt/nmtPreInit.hpp [hunk 14]
```diff
@@ -23,19 +23,19 @@
...
-#include "runtime/atomic.hpp"
-#endif
-#include "services/memTracker.hpp"
-#include "utilities/checkedCast.hpp"
+#include "nmt/memTracker.hpp"
 #include "utilities/debug.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/macros.hpp"
 
+#ifdef ASSERT
```
Comment (stefank) [inline src/hotspot/share/nmt/nmtPreInit.hpp:35]: The blank line at 34 is not following the style for our conditional includes. Remove it, or better yet skip conventionalize the include of runtime/atomic.hpp since it just adds to noise to the file.

**Review by stefank — CHANGES_REQUESTED** [review]

File: test/hotspot/gtest/nmt/test_nmt_cornercases.cpp [hunk 15]
```diff
@@ -24,10 +24,11 @@
 
 #include "precompiled.hpp"
 #include "memory/allocation.hpp"
+#include "nmt/mallocHeader.inline.hpp"
+#include "nmt/mallocTracker.hpp"
 #include "runtime/os.hpp"
-#include "services/mallocHeader.inline.hpp"
-#include "services/mallocTracker.hpp"
-#include "services/memTracker.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline test/hotspot/gtest/nmt/test_nmt_cornercases.cpp:30]: Sort order

File: src/hotspot/share/services/mallocLimit.cpp [hunk 16]
```diff
@@ -29,7 +29,7 @@
 #include "runtime/java.hpp"
 #include "runtime/globals.hpp"
 #include "services/mallocLimit.hpp"
-#include "services/nmtCommon.hpp"
+#include "nmt/nmtCommon.hpp"
```
Comment (stefank) [inline src/hotspot/share/services/mallocLimit.cpp:32]: Sort order

**Comment by stefank** [comment]
> For the gtest source files I separated the includes in a consistent manner, they all look like this pattern now:

That's not what I see in the latest patch. Could you revert that separation and then we can consider that style change in a separate RFE?

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/share/nmt/nmtPreInit.hpp (hunk 14 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/nmtPreInit.hpp:35]: Removed the blank line, I think the noise is meaningful in this case.

**Comment by jdksjolen** [comment]
> > For the gtest source files I separated the includes in a consistent manner, they all look like this pattern now:
> 
> That's not what I see in the latest patch. Could you revert that separation and then we can consider that style change in a separate RFE?

Sure, this could be done for all tests (not only NMT).

**Comment by jdksjolen** [comment]
@stefank, I believe that I've applied your suggestions, would you mind having another look? Thanks

File: test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp [hunk 17]
```diff
@@ -24,14 +24,14 @@
...
 #include "runtime/os.hpp"
-#include "services/mallocHeader.inline.hpp"
-#include "services/memTracker.hpp"
+#include "testutils.hpp"
 #include "unittest.hpp"
 
 // Uncomment to get test output
 //#define LOG_PLEASE
-#include "testutils.hpp"
```
Comment (stefank) [inline test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp:35]: There seems to be an ordering requirement between LOG_PLEASE and the inclusion of LOG_PLEASE. I think you break it with this change.

File: test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp [hunk 18]
```diff
@@ -24,12 +24,12 @@
...
 #include "precompiled.hpp"
 #include "memory/allocation.hpp"
+#include "nmt/memTracker.hpp"
 #include "runtime/os.hpp"
-#include "services/memTracker.hpp"
 #include "utilities/debug.hpp"
 #include "utilities/ostream.hpp"
-#include "unittest.hpp"
 #include "testutils.hpp"
+#include "unittest.hpp"
```
Comment (stefank) [inline test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp:33]: It is inclear to me if there is an ordering requirement between testutils and unittest. I'd prefer if you didn't change those unit test header files in this patch and make these cleanups separately.

File: src/hotspot/share/nmt/mallocHeader.cpp [hunk 19]
```diff
@@ -24,15 +24,14 @@
  */
 #include "precompiled.hpp"
 
-#include "services/mallocHeader.inline.hpp"
+#include "nmt/mallocHeader.inline.hpp"
 
+#include "nmt/mallocSiteTable.hpp"
```
Comment (stefank) [inline src/hotspot/share/nmt/mallocHeader.cpp:29]: There should be no newline between these two includes. If you make another round of changes I think it would be good to get rid of it.

File: src/hotspot/share/memory/resourceArea.inline.hpp [hunk 20]
```diff
@@ -27,7 +27,7 @@
 
 #include "memory/resourceArea.hpp"
 
-#include "services/memTracker.hpp"
+#include "nmt/memTracker.hpp"
```
Comment (stefank) [inline src/hotspot/share/memory/resourceArea.inline.hpp:30]: Another new line that should not be here. Bonus points if it gets removed.

**Review by stefank — APPROVED** [review]
I've approved the patch, but I think you should revert the two gtest changes I mention below. I have given suggestions for two cleanups that you might want to do.

**Comment by jdksjolen** [comment]
Thanks Stefan, good catch on the ordering requirements. I'll integrate this ASAP.

**Review by stefank — APPROVED** [review]

**Comment by jdksjolen** [comment]
/integrate

Thank you for the reviews!
